harness/api_checkers.go

```diff
@@ -0,0 +1,437 @@
+	if !assert.NotNil(t, spanContext, "assertEmptyBaggage got empty context") {
+		return
+	}
+	spanContext.ForeachBaggageItem(func(k, v string) bool {
+		assert.Fail(t, "new span shouldn't have baggage")
+		return false
+	})
+}
+
+// TestSpanBaggage tests calls to set and get span baggage, and if the CheckBaggageValues option
+// is set, asserts that baggage values were successfully retrieved.
+func (s *APICheckSuite) TestSpanBaggage() {
+	span := s.tracer.StartSpan("Fry")
+	assertEmptyBaggage(s.T(), span.Context())
+
+	spanRef := span.SetBaggageItem("Kiff-loves", "Amy")
+	assert.Exactly(s.T(), spanRef, span)
+
+	val := span.BaggageItem("Kiff-loves")
+	if s.opts.CheckBaggageValues {
+		assert.Equal(s.T(), "Amy", val)
+	} else {
+		s.T().Log("Baggage propagation not supported, not checking")
+	}
+	span.Finish()
+}
+
+// TestContextBaggage tests calls to set and get span baggage, and if the CheckBaggageValues option
+// is set, asserts that baggage values were successfully retrieved from the span's SpanContext.
+func (s *APICheckSuite) TestContextBaggage() {
+	span := s.tracer.StartSpan("Fry")
+	assertEmptyBaggage(s.T(), span.Context())
+
+	span.SetBaggageItem("Kiff-loves", "Amy")
+	if s.opts.CheckBaggageValues {
+		called := false
+		span.Context().ForeachBaggageItem(func(k, v string) bool {
+			assert.False(s.T(), called)
+			called = true
+			assert.Equal(s.T(), "Kiff-loves", k)
+			assert.Equal(s.T(), "Amy", v)
+			return true
+		})
+	} else {
+		s.T().Log("Baggage propagation not supported, not checking")
+	}
+	span.Finish()
+}
+
+// TestTextPropagation tests if the Tracer can Inject a span into a TextMapCarrier, and later Extract it.
+// If CheckExtract is set, it will check if Extract was successful (returned no error). If a Probe is set,
+// it will check if the extracted context is in the same trace as the original span.
+func (s *APICheckSuite) TestTextPropagation() {
+	span := s.tracer.StartSpan("Bender")
+	textCarrier := opentracing.TextMapCarrier{}
+	err := span.Tracer().Inject(span.Context(), opentracing.TextMap, textCarrier)
+	assert.NoError(s.T(), err)
+
+	extractedContext, err := s.tracer.Extract(opentracing.TextMap, textCarrier)
+	if s.opts.CheckExtract {
+		assert.NoError(s.T(), err)
+		assertEmptyBaggage(s.T(), extractedContext)
+	} else {
+		s.T().Log("Tracer.Extract not supported, not checking")
+	}
+	if s.opts.Probe != nil {
+		assert.True(s.T(), s.opts.Probe.SameSpanContext(span, extractedContext))
+	}
+	span.Finish()
+}
+
+// TestHTTPPropagation tests if the Tracer can Inject a span into HTTP headers, and later Extract it.
+// If CheckExtract is set, it will check if Extract was successful (returned no error). If a Probe is set,
+// it will check if the extracted context is in the same trace as the original span.
+func (s *APICheckSuite) TestHTTPPropagation() {
+	span := s.tracer.StartSpan("Bender")
+	textCarrier := opentracing.HTTPHeadersCarrier{}
+	// XXX add same test cases around valid HTTP header characters, casing
```

Comment:
btw, the comment is especially relevant to baggage, because it had different requirements in the past (allowed more limitations), and not all tracers support it correctly. I would make that "strict" check controllable via capabilities flag.